GoogleDataTransport/GDTCCTLibrary/GDTCCTUploadOperation.m

```diff
@@ -185,99 +183,81 @@ - (void)uploadTarget:(GDTCORTarget)target withConditions:(GDTCORUploadConditions
-                [self uploadBatchWithID:batchID
-                                 events:events
-                                 target:target
-                                storage:storage
-                             completion:handler];
-              }];
-}
-
-/** Performs URL request, handles the result and updates the uploader state. */
-- (void)uploadBatchWithID:(nullable NSNumber *)batchID
-                   events:(nullable NSSet<GDTCOREvent *> *)events
-                   target:(GDTCORTarget)target
-                  storage:(id<GDTCORStorageProtocol>)storage
-               completion:(dispatch_block_t)completion {
-  [self
-      sendURLRequestForBatchWithID:batchID
-                            events:events
-                            target:target
-                 completionHandler:^(NSNumber *_Nonnull batchID,
-                                     NSSet<GDTCOREvent *> *_Nullable events, NSData *_Nullable data,
-                                     NSURLResponse *_Nullable response, NSError *_Nullable error) {
-                   dispatch_async(self.uploaderQueue, ^{
-                     [self handleURLResponse:response
-                                        data:data
-                                       error:error
-                                      target:target
-                                     storage:storage
-                                     batchID:batchID];
-#if !NDEBUG
-                     // Post a notification when in DEBUG mode to state how many packages
-                     // were uploaded. Useful for validation during tests.
-                     [[NSNotificationCenter defaultCenter]
-                         postNotificationName:GDTCCTUploadCompleteNotification
-                                       object:@(events.count)];
-#endif  // #if !NDEBUG
-                     self.isCurrentlyUploading = NO;
-                     completion();
-                   });
-                 }];
+/** Sends URL request for to upload the provided batch and handles the response. */
+- (FBLPromise<NSNull *> *)sendURLRequestWithBatch:(GDTCORUploadBatch *)batch
+                                           target:(GDTCORTarget)target
+                                          storage:(id<GDTCORStoragePromiseProtocol>)storage {
+  NSNumber *batchID = batch.batchID;
+
+  // 1. Send URL request.
+  return [self sendURLRequestWithBatch:batch target:target]
+      .thenOn(
+          self.uploaderQueue,
+          ^FBLPromise<NSNull *> *(GULURLSessionDataResponse *response) {
+            // 2. Parse response and update the next upload time if can.
+            [self updateNextUploadTimeWithResponse:response forTarget:target];
+
+            // 3. Cleanup batch.
+
+            // Only retry if one of these codes is returned.
+            NSInteger statusCode = response.HTTPResponse.statusCode;
+            if (statusCode == 429 || statusCode == 503) {
```

Comment:
Please add comment about 429 and 503 definition

Reply:
added